scheduler.rst

```diff
@@ -0,0 +1,328 @@
+
+    Also, `crontab_helper`_ is a good tool if you need help to construct/understand cron expressions
+
+.. versionadded:: 6.4
+
+    Since version 6.4, it is now possible to add and define a timezone as a 3rd argument
+
+#. It can be a periodical trigger through various frequency formats (string / integer / DateInterval)
+
+.. configuration-block::
+
+    .. code-block:: php
+
+        RecurringMessage::every('10 seconds', new Message());
+        RecurringMessage::every('3 weeks', new Message());
+        RecurringMessage::every('first Monday of next month', new Message());
+
+        $from = new \DateTimeImmutable('13:47', new \DateTimeZone('Europe/Paris'));
+        $until = '2023-06-12';
+        RecurringMessage::every('first Monday of next month', new Message(), $from, $until);
+
+#. It can be a custom trigger implementing :class:`Symfony\\Component\\Scheduler\\TriggerInterface`
+
+If you go back to your scenario regarding reports generation based on your customer preferences.
+If the basic frequency is set to a daily basis, you will need to implement a custom trigger due to the specific requirement of not generating reports during public holiday periods::
+
+    // src/Scheduler/Trigger/NewUserWelcomeEmailHandler.php
+    namespace App\Scheduler\Trigger;
+
+    class ExcludeHolidaysTrigger implements TriggerInterface
+    {
+        public function __construct(private TriggerInterface $inner)
+        {
+        }
+
+        public function __toString(): string
+        {
+            return $this->inner.' (except holidays)';
+        }
+
+        public function getNextRunDate(\DateTimeImmutable $run): ?\DateTimeImmutable
+        {
+            if (!$nextRun = $this->inner->getNextRunDate($run)) {
+                return null;
+            }
+
+            while (!$this->isHoliday($nextRun) { // loop until you get the next run date that is not a holiday
+                $nextRun = $this->inner->getNextRunDate($nextRun);
+            }
+
+            return $nextRun;
+        }
+
+        private function isHoliday(\DateTimeImmutable $timestamp): bool
+        {
+            // app specific logic to determine if $timestamp is on a holiday
+            // returns true if holiday, false otherwise
+        }
+    }
+
+Then, you would have to define your RecurringMessage
+
+.. configuration-block::
+
+    .. code-block:: php
+
+        RecurringMessage::trigger(
+            new ExcludeHolidaysTrigger( // your custom trigger wrapper
+                CronExpressionTrigger::fromSpec('@daily'),
+            ),
+            new SendDailySalesReports(),
+        );
+
+The RecurringMessages must be attached to a Schedule::
+
+    // src/Scheduler/MyScheduleProvider.php
+    namespace App\Scheduler;
+
+    #[AsSchedule('uptoyou')]
+    class MyScheduleProvider implements ScheduleProviderInterface
+    {
+        public function getSchedule(): Schedule
+        {
+            return $this->schedule ??= (new Schedule())
```

Comment:
Is the `$this->schedule` part of the memoization? I really need to be convinced that it is VERY important to keep this detail